pkg/skaffold/walk/walk.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package walk
+
+import (
+	"os"
+
+	"github.com/karrick/godirwalk"
+)
+
+// Dirent stores the name and type of a file system entry.
+type Dirent interface {
+	IsDir() bool
+	Name() string
+}
+
+// Predicate represents a predicate on file system entries.
+// Given a file's absolute path and information, it returns `true`
+// when the predicate is matched. It can also return a `filepath.SkipDir`
+// error to skip a directory and its children altogether.
+type Predicate func(absPath string, info Dirent) (bool, error)
+
+// Action is a function that takes a file's absolute path and information,
+// and optionally returns an error.
+type Action func(absPath string, info Dirent) error
+
+type Builder interface {
+	// Options
+	Unsorted() Builder
+
+	// Predicates
+	When(Predicate) Builder
+	WhenIsDir() Builder
+	WhenIsFile() Builder
+	WhenHasName(string) Builder
+
+	// Actions
+	Do(Action) error
+	MustDo(Action)
+	AppendPaths(*[]string) error
+	CollectPaths() ([]string, error)
+}
+
+type builder struct {
+	dir       string
+	unsorted  bool
+	predicate Predicate
+}
+
+func From(dir string) Builder {
+	return &builder{
+		dir:       dir,
+		unsorted:  false,
+		predicate: func(string, Dirent) (bool, error) { return true, nil },
+	}
+}
+
+func (w *builder) Unsorted() Builder {
+	w.unsorted = true
+	return w
+}
+
+func (w *builder) When(predicate Predicate) Builder {
+	w.predicate = and(w.predicate, predicate)
+	return w
+}
+
+func (w *builder) WhenIsFile() Builder {
+	return w.When(isFile)
+}
+
+func (w *builder) WhenIsDir() Builder {
+	return w.When(isDir)
+}
+
+func (w *builder) WhenHasName(name string) Builder {
+	return w.When(hasName(name))
+}
+
+func (w *builder) AppendPaths(values *[]string) error {
+	return w.Do(appendPaths(values))
+}
+
+func (w *builder) CollectPaths() ([]string, error) {
+	var paths []string
+	err := w.Do(appendPaths(&paths))
+	return paths, err
+}
+
+func (w *builder) Do(action Action) error {
+	info, err := os.Lstat(w.dir)
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		match, err := w.predicate(w.dir, info)
+		if !match || err != nil {
+			return err
+		}
+
+		return action(w.dir, info)
+	}
+
+	return godirwalk.Walk(w.dir, &godirwalk.Options{
+		Unsorted: w.unsorted,
+		Callback: func(path string, info *godirwalk.Dirent) error {
+			match, err := w.predicate(path, info)
```

Comment:
It seems most of the uses of `Walk()` relativize the path against the dir.  I wonder if it would make sense to relativize the path here and pass in the relative path into the predicate and action.  Or maybe it can be controlled by an option on the builder? `walk.From(xxx).RelativePaths().When(yyy).Do(zzz)`.

Reply:
Yeah, I've seen that too. Maybe in a next PR though?